mapiproxy/servers/default/emsmdb/emsmdbp_object.c

```diff
@@ -3129,3 +3132,221 @@ _PUBLIC_ struct emsmdbp_object *emsmdbp_object_ftcontext_init(TALLOC_CTX *mem_ct
+	char			       *xml;
+	struct emsmdbp_prop_index      prop_index;
+	int			       i;
+	enum MAPISTATUS		       *retvals = NULL;
+	struct mapistore_message       *msg_data = NULL;
+	const enum MAPITAGS	       sharing_prop_tags[] = {PidLidSharingLocalType, PidLidSharingInitiatorName, PidLidSharingInitiatorSmtp, PidLidSharingInitiatorEntryId, PidLidSharingRemoteName, PidLidSharingRemoteUid, PidLidSharingRemoteStoreUid}; /* Only managing invitations by now */
+	struct SBinary_short	       *entryId;
+	struct SPropTagArray	       *sharing_properties;
+	struct SPropValue	       *attach_bin_property;
+	TALLOC_CTX		       *local_mem_ctx;
+	const uint32_t		       SHARING_PROPS_COUNT = sizeof(sharing_prop_tags) / sizeof(enum MAPITAGS);
+	uint32_t		       contextID;
+	void			       **data_pointers;
+
+	attach_bin_property = talloc_zero(mem_ctx, struct SPropValue);
+	OPENCHANGE_RETVAL_IF(!attach_bin_property, MAPI_E_NOT_ENOUGH_MEMORY, NULL);
+
+	local_mem_ctx = talloc_named(NULL, 0, "sharing_metadata_property");
+	OPENCHANGE_RETVAL_IF(!local_mem_ctx, MAPI_E_NOT_ENOUGH_MEMORY, NULL);
+
+	/* We get the required properties from the mapistore */
+	sharing_properties = talloc_zero(mem_ctx, struct SPropTagArray);
+	OPENCHANGE_RETVAL_IF(!sharing_properties, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+	sharing_properties->cValues = SHARING_PROPS_COUNT;
+	sharing_properties->aulPropTag = talloc_array(local_mem_ctx, enum MAPITAGS, SHARING_PROPS_COUNT);
+	memcpy(sharing_properties->aulPropTag, sharing_prop_tags, sizeof(sharing_prop_tags));
+
+	data_pointers = emsmdbp_object_get_properties(local_mem_ctx, emsmdbp_ctx, sharing_object,
+						      sharing_properties, &retvals);
+	OPENCHANGE_RETVAL_IF(!data_pointers || retvals[0] != MAPI_E_SUCCESS, MAPI_E_NOT_FOUND, local_mem_ctx);
+
+	/* TODO: We do only support calendar invitation by now. Checks on PidLidSharingLocalType */
+	if (strncmp((char *)data_pointers[0], "IPF.Appointment", strlen("IPF.Appointment")) != 0) {
+		talloc_free(local_mem_ctx);
+		return MAPI_E_NOT_FOUND;
+	}
+
+	/* Check every retval is SUCCESS to build up the XML */
+	for (i = 1; i < SHARING_PROPS_COUNT; i++) {
+		OPENCHANGE_RETVAL_IF(retvals[i] != MAPI_E_SUCCESS, MAPI_E_NOT_FOUND, local_mem_ctx);
+	}
+
+	xml = talloc_asprintf(local_mem_ctx,
+			      "<?xml version=\"1.0\"?> "
+			      "<SharingMessage xmlns=\"http://schemas.microsoft.com/sharing/2008\">"
+			      "<DataType>calendar</DataType>"
+			      "<Initiator>");
+	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+	/*** Initiator ***/
+	xml = talloc_asprintf_append(xml, "<Name>%s</Name>", (char *)data_pointers[1]);	 /* PidLidSharingInitiatorName */
+	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+	xml = talloc_asprintf_append(xml, "<SmtpAddress>%s</SmtpAddress>", (char *)data_pointers[2]); /*PidLidSharingInitiatorSmtp */
+	OPENCHANGE_RETVAL_IF(!xml, MAPI_E_NOT_ENOUGH_MEMORY, local_mem_ctx);
+
+	if (retvals[2] == MAPI_E_SUCCESS) {
```

Comment:
Is this necessary? We have already checked that every retvals[i] is `MAPI_E_SUCCESS` at line 3177